src/AutoRest.CSharp/MgmtTest/Generation/ScenarioTestWriter.cs

```diff
@@ -0,0 +1,259 @@
+
+namespace AutoRest.CSharp.MgmtTest.Generation
+{
+    /// <summary>
+    /// Code writer for resource test.
+    /// </summary>
+    internal class ScenarioTestWriter: ClientWriter
+    {
+        private TestDefinitionModel _testDef;
+        private CodeWriter _writer;
+        protected string TestNamespace => MgmtContext.Context.DefaultNamespace + ".Tests.Scenario";
+        public string TypeName => System.IO.Path.GetFileNameWithoutExtension(_testDef.FilePath).ToCleanName();
+        protected string TestBaseName => $"ManagementRecordedTestBase<{MgmtContext.Context.DefaultLibraryName}TestEnvironment>";
+
+        public ScenarioTestWriter(CodeWriter writer, TestDefinitionModel testDef)
+        {
+            this._testDef = testDef;
+            this._writer = writer;
+        }
+
+        public void WriteScenarioTest()
+        {
+            WriteUsings(_writer);
+
+            using (_writer.Namespace(TestNamespace))
+            {
+                _writer.WriteXmlDocumentationSummary($"Test generated from Api Scenario {_testDef.FilePath}");
+                _writer.Append($"public partial class {TypeName:D} : ");
+                _writer.Line($"{TestBaseName}");
+                using (_writer.Scope())
+                {
+                    WriteTesterCtors();
+
+                    WritePrepareMethod();
+                    WriteCleanUpMethod();
+
+                    foreach (var scenario in _testDef.Scenarios)
+                    {
+                        WriteScenarioMethod(scenario);
+                    }
+                }
+            }
+        }
+
+        protected void WriteUsings(CodeWriter writer)
+        {
+            writer.UseNamespace("Azure.Core.TestFramework");
+            writer.UseNamespace("Azure.ResourceManager.TestFramework");
+        }
+
+        protected void WritePrepareMethod()
+        {
+            // TODO: generate prepare step
+        }
+
+        protected void WriteCleanUpMethod()
+        {
+            // TODO: generate cleanUp step
+        }
+
+        protected IEnumerable<string> GetAllScenarioDefinedVariables(Scenario? scenario, TestStep step)
+        {
+            HashSet<string> scenarioDefinedVariables = _testDef.GetAllVariableNames();
+            if (scenario is not null)
+            {
+                scenarioDefinedVariables.UnionWith(scenario.GetAllVariableNames());
+            }
+            scenarioDefinedVariables.UnionWith(step.GetAllVariableNames());
+            return scenarioDefinedVariables;
+        }
+
+        protected void WriteStep(Scenario? scenario, TestStep step)
+        {
+            _writer.Line($"// Step: {step.Step}");
+            var existedVariables = _testDef.GetAllVariableNames();
+            if (scenario is not null)
+            {
+                existedVariables.UnionWith(scenario.GetAllVariableNames());
+            }
+            using (new CodeWriterOavVariableScope(_writer, step, existedVariables))
```

Comment:
What does this scope mean?

Reply:
it is for handling variables with same name occurred in oav def/scenario/step level. For instance, the variable 'v1' defined in both scenario and inside step but has different value.
~~~
scenario:
  variables:
     v1: foo
  steps:
    - step: step1
      variables:
        v1: bar
~~~
CodeWriterOavVariableScope is for back up 'v1' in front of step1, and restore 'v1' after step1 is done